src/Storages/MergeTree/registerStorageMergeTree.cpp

```diff
@@ -480,50 +482,70 @@ static StoragePtr create(const StorageFactory::Arguments & args)
                     ErrorCodes::BAD_ARGUMENTS);
 
         --arg_cnt;
+        merging_param_key_arg = std::make_shared<ASTIdentifier>(merging_params.version_column);
```

Comment:
Maybe it's possible to copy sorting_key and add this column to it? Because now passing `additional_column` parameter because of only one use-case looks not very convinient.

Reply:
I added three helper methods https://github.com/ClickHouse/ClickHouse/pull/11516/files#diff-8226568d27f81979c04e8f6214baca3eR40-R64 for keys. Also add a check, that we don't lose additional_column during assignment https://github.com/ClickHouse/ClickHouse/pull/11516/files#diff-987470f0bbd6c19bd6bdc9cd5ed59811R48-R49.  Also add a lot of comments in registerStorageMergeTree https://github.com/ClickHouse/ClickHouse/pull/11516/files#diff-b72871a5204b0c19d551305c5b854544R519 and description of additional_column fields https://github.com/ClickHouse/ClickHouse/pull/11516/files#diff-8226568d27f81979c04e8f6214baca3eR33-R36. 

To be honest, keys still seem slightly confusing even for me... But I think the main reason for this is the too heterogeneous meaning and way of working with them in the merge tree:

1. Partition key -- not required in storage definition, but must present as empty expressions list, empty columns, and sample block. It cannot be changed in alter.
2. Sorting key -- required in storage definition (at least empty `ORDER BY tuple()`), may have an additional column depending on storage type (but not in the definition), may be changed in alter.
3. Primary key -- not required in storage definition, but must present as a copy of sorting key without additional column and definition. It cannot be changed in alter but can be copied from sorting key (definition will appear, but additional column shouldn't).
4. Sampling key -- simplest one, not required, may be absent (both definition and expressions). It cannot be altered.

The ultimate way for sorting the key is to create a separate structure with an additional column and duplicate some code. But I don't sure that the code will become significantly more clear.